cmd/registry/cmd/compute/complexity.go

```diff
@@ -109,57 +110,64 @@ func (task *computeComplexityTask) String() string {
 }
 
 func (task *computeComplexityTask) Run(ctx context.Context) error {
-	request := &rpc.GetApiSpecContentsRequest{
-		Name: task.specName,
-	}
-	spec, err := task.client.GetApiSpecContents(ctx, request)
+	specName, err := names.ParseSpecRevision(task.specName)
 	if err != nil {
 		return err
 	}
-	relation := "complexity"
-	log.Debugf(ctx, "Computing %s/artifacts/%s", task.specName, relation)
-	var complexity *metrics.Complexity
-	if core.IsOpenAPIv2(spec.GetContentType()) {
-		document, err := oas2.ParseDocument(spec.GetData())
-		if err != nil {
-			log.FromContext(ctx).WithError(err).Errorf("Invalid OpenAPI: %s", task.specName)
-			return nil
-		}
-		complexity = core.SummarizeOpenAPIv2Document(document)
-	} else if core.IsOpenAPIv3(spec.GetContentType()) {
-		document, err := oas3.ParseDocument(spec.GetData())
-		if err != nil {
-			log.FromContext(ctx).WithError(err).Errorf("Invalid OpenAPI: %s", task.specName)
-			return nil
-		}
-		complexity = core.SummarizeOpenAPIv3Document(document)
-	} else if core.IsDiscovery(spec.GetContentType()) {
-		document, err := discovery.ParseDocument(spec.GetData())
-		if err != nil {
-			log.FromContext(ctx).WithError(err).Errorf("Invalid Discovery: %s", task.specName)
-			return nil
-		}
-		complexity = core.SummarizeDiscoveryDocument(document)
-	} else if core.IsProto(spec.GetContentType()) && core.IsZipArchive(spec.GetContentType()) {
-		complexity, err = core.SummarizeZippedProtos(spec.GetData())
-		if err != nil {
-			log.FromContext(ctx).WithError(err).Errorf("Error processing protos: %s", task.specName)
-			return nil
-		}
-	} else {
-		return fmt.Errorf("we don't know how to summarize %s", task.specName)
-	}
+	return core.GetSpecRevision(ctx, task.client, specName, true,
```

Comment:
Good point - I noticed that but wanted to make a change that was most consistent with the existing API.

This change would have been simpler if we had not made [this change](https://github.com/apigee/registry/commit/8ab7da420f82f37a59ab218ce8aa5c931f2cfa3d#diff-a7a1d9fa477282d8e5d6308097b87cc450729521a5935287ab5e4c941adf8a52R103), which removed the requested items from the return values of the `Get` commands. 

I've restructured the code to minimize the line count affected by this change, and if needed, we can separately discuss reversion of the previous one.

Reply:
Having 99% of callsites ignore the return value and this particular one use it instead of the callback seems like the worst of all scenarios to me, so I'm glad you didn't have that option here. That change eliminated the "two ways of using these functions" problem.

Now we can decide whether to stick with callbacks or move entirely to return values. Ideally I think the requested items should always be returned instead of forwarded to a callback, but it's questionable whether that's worth the effort.

Reply:
I'm not sure where the "99%" number came from (is it rhetorical?). Since there are places in our existing code in which both access methods are useful, insisting on this style guide rule will lead us to having two getters for each resource, probably with the callback version wrapping the direct one.